phd-tests/framework/src/test_vm/mod.rs

```diff
@@ -591,49 +604,49 @@ impl TestVm {
-                async move {
-                    for step in boot_sequence.0 {
-                        debug!(?step, "executing command in boot sequence");
-                        match step {
-                            CommandSequenceEntry::WaitFor(s) => {
-                                self.wait_for_serial_output_async(
-                                    s,
-                                    Duration::MAX,
-                                )
+        match timeout(
+            timeout_duration,
+            async move {
+                for step in boot_sequence.0 {
+                    debug!(?step, "executing command in boot sequence");
+                    match step {
+                        CommandSequenceEntry::WaitFor(s) => {
+                            self.wait_for_serial_output(s, Duration::MAX)
                                 .await?;
-                            }
-                            CommandSequenceEntry::WriteStr(s) => {
-                                self.send_serial_str_async(s).await?;
-                                self.send_serial_str_async("\n").await?;
-                            }
-                            CommandSequenceEntry::ChangeSerialConsoleBuffer(
-                                kind,
-                            ) => {
-                                self.change_serial_buffer_kind(kind)?;
-                            }
-                            CommandSequenceEntry::SetSerialByteWriteDelay(
-                                duration,
-                            ) => {
-                                self.set_serial_byte_write_delay(duration)?;
-                            }
+                        }
+                        CommandSequenceEntry::WriteStr(s) => {
+                            self.send_serial_str(s).await?;
+                            self.send_serial_str("\n").await?;
+                        }
+                        CommandSequenceEntry::ChangeSerialConsoleBuffer(
+                            kind,
+                        ) => {
+                            self.change_serial_buffer_kind(kind)?;
+                        }
+                        CommandSequenceEntry::SetSerialByteWriteDelay(
+                            duration,
+                        ) => {
+                            self.set_serial_byte_write_delay(duration)?;
                         }
                     }
-                    Ok::<(), anyhow::Error>(())
                 }
-                .instrument(info_span!("wait_to_boot")),
-            )
-            .await
-            .map_err(|e| anyhow!(e))
-        })?;
+                Ok::<(), anyhow::Error>(())
+            }
+            .instrument(info_span!("wait_to_boot")),
+        )
+        .await
+        {
+            Err(_) => anyhow::bail!("timed out while waiting to boot"),
+            Ok(inner) => {
+                inner.context("executing guest login sequence")?;
+            }
+        };
```

Comment:
style nit, take it or leave it: this is a _huge_ amount of code for the expression in a `match`, and IMO, it's easy to forget that this is a match while reading the code in the expression part. what do you think of something more like this:
```rust
        let wait_to_boot = async move {
            for step in boot_sequence.0 {
                debug!(?step, "executing command in boot sequence");
                match step {
                    CommandSequenceEntry::WaitFor(s) => {
                        self.wait_for_serial_output(s, Duration::MAX)
                            .await?;
                    }
                    CommandSequenceEntry::WriteStr(s) => {
                        self.send_serial_str(s).await?;
                        self.send_serial_str("\n").await?;
                    }
                    CommandSequenceEntry::ChangeSerialConsoleBuffer(
                        kind,
                    ) => {
                        self.change_serial_buffer_kind(kind)?;
                    }
                    CommandSequenceEntry::SetSerialByteWriteDelay(
                        duration,
                    ) => {
                        self.set_serial_byte_write_delay(duration)?;
                    }
                }
            }
            Ok::<(), anyhow::Error>(())
        }
        .instrument(info_span!("wait_to_boot"));
        match timeout(timeout_duration, wait_to_boot).await {
            Err(_) => anyhow::bail!("timed out while waiting to boot"),
            Ok(inner) => {
                inner.context("executing guest login sequence")?;
            }
        };
```